### PR TITLE
Support Hibernate 3.5

### PR DESCRIPTION
**Pull request by willgittoes-dd**
Branches off #765 

Adds support for Hibernate 3.5. Mostly a copy of the v4 impl, except that there is no unifying interface of SharedSessionContract between Session and StatelessSession, so there's a lot of 'instanceof' and duplication. I couldn't get a nicely polymorphic solution, because I can't find a way to work around the limitations of the Advice pattern. 

I had a version that used the same instrumentation code for v3 and v4 (making the 2-way 'instanceof' if-chain into a 3-way one for Session, StatelessSession, and SharedSessionContract). But it was uglier in practice.

### CODE REVIEW

**Comment by tylerbenson** [comment]
Could you add a comment/note/readme somewhere explaining the difference you stated above between this and the v4 instrumentation?  I would have also been ok with extending the v4 instrumentation using the 3 way check as you mentioned, but I think this is also a good solution.  Go ahead and merge #765 and rebase and I'll take another look, though at a cursory glance it seems ok to me.

**Comment by willgittoes-dd** [comment]
I ended up moving the 'common' hibernate stuff to its own module and package. Muzzle tests weren't passing because the v4 instrumentation that I was pulling in just for SessionState & co had some classes that referenced Hibernate classes that aren't in v3. And even though none of that code was ever executed (nor the classes even used) by the v3 integration, I was getting ClassDefNotFound in the muzzle tests (but not the unit tests).

File: dd-java-agent/instrumentation/hibernate-common/src/main/java/datadog/trace/instrumentation/hibernate/common/SessionMethodUtils.java [hunk 1]
```diff
@@ -43,8 +41,8 @@
               .buildSpan(operationName)
               .asChildOf(sessionState.getSessionSpan())
               .startActive(true);
-      DECORATOR.afterStart(scope.span());
-      DECORATOR.onOperation(scope.span(), entity);
+      HibernateDecorator.DECORATOR.afterStart(scope.span());
+      HibernateDecorator.DECORATOR.onOperation(scope.span(), entity);
```
[the remark this answers is not in the document]
Comment (willgittoes-dd) [reply dd-java-agent/instrumentation/hibernate-common/src/main/java/datadog/trace/instrumentation/hibernate/common/SessionMethodUtils.java]: Yeah, the IDE just "helpfully" changed it for me, I've changed it back.

File: settings.gradle [hunk 2]
```diff
@@ -40,8 +40,10 @@ include ':dd-java-agent:instrumentation:elasticsearch:transport-5'
 include ':dd-java-agent:instrumentation:elasticsearch:transport-5.3'
 include ':dd-java-agent:instrumentation:elasticsearch:transport-6'
 include ':dd-java-agent:instrumentation:grpc-1.5'
+include ':dd-java-agent:instrumentation:hibernate-3.5'
 include ':dd-java-agent:instrumentation:hibernate-4.0'
 include ':dd-java-agent:instrumentation:hibernate-4.3'
+include ':dd-java-agent:instrumentation:hibernate-common'
```
[the remark this answers is not in the document]
Comment (willgittoes-dd) [reply settings.gradle]: I did this, although I named the packages like '..hibernate.core.v3_5'.

**Pull request merged**